src/gradient.zig

```diff
@@ -0,0 +1,587 @@
+    /// stops that have been added using `Stop.List.add`.
+    pub fn init(x0: f64, y0: f64, x1: f64, y1: f64) Linear {
+        return .{
+            .start = .{ .x = x0, .y = y0 },
+            .end = .{ .x = x1, .y = y1 },
+            .stops = .{},
+        };
+    }
+
+    /// Initializes the gradient with externally allocated memory for the
+    /// stops. Do not use this with `deinit` or `Stop.List.add` as it will
+    /// cause illegal behavior, use `Stop.List.addAssumeCapacity` instead.
+    pub fn initBuffer(x0: f64, y0: f64, x1: f64, y1: f64, stops: []Stop) Linear {
+        return .{
+            .start = .{ .x = x0, .y = y0 },
+            .end = .{ .x = x1, .y = y1 },
+            .stops = .{ .l = std.ArrayListUnmanaged(Stop).initBuffer(stops) },
+        };
+    }
+
+    /// Releases any stops that have been added using `Stop.List.add`. Must use
+    /// the same allocator that was used there.
+    pub fn deinit(self: *Linear, alloc: mem.Allocator) void {
+        self.stops.deinit(alloc);
+    }
+
+    /// Returns this gradient as a pattern.
+    pub fn asPatternInterface(self: *const Linear) Pattern {
+        return .{ .linear_gradient = self };
+    }
+
+    /// Gets the pixel calculated for the gradient at `(x, y)`.
+    pub fn getPixel(self: *const Linear, x: i32, y: i32) pixel.Pixel {
+        return self.stops.search(self.getOffset(x, y)).lerp().asPixel();
+    }
+
+    /// Performs orthogonal projection on the gradient, transforming the
+    /// supplied (x, y) co-ordinates into an offset. This offset can be used to
+    /// manually search on the gradient's stops using `Stop.List.search`.
+    ///
+    /// ```
+    /// const offset = gradient.getOffset(50, 50);
+    /// const result = gradient.stops.search(offset);
+    /// ... // (lerp off of search result or perform other operations)
+    /// ```
+    pub fn getOffset(self: *const Linear, x: i32, y: i32) f64 {
+        return project(self.start, self.end, .{
+            .x = @as(f64, @floatFromInt(x)),
+            .y = @as(f64, @floatFromInt(y)),
+        });
+    }
+};
+
+/// Represents a color stop in a gradient.
+///
+/// Color stops within gradients are always stored as RGBA, they will be
+/// converted according to the surface they are applied to.
+pub const Stop = struct {
+    idx: usize,
+
+    /// The color for this color stop.
+    color: pixel.RGBA,
+
+    /// The offset of this color stop, clamped between `0.0` and `1.0`.
+    offset: f64,
+
+    /// Represents a list of color stops. Do not copy this field directly from
+    /// gradient to gradient as it may cause the index to go out of sync.
+    const List = struct {
+        current_idx: usize = 0,
+        l: std.ArrayListUnmanaged(Stop) = .{},
+
+        /// Initializes a `List` with externally allocated memory. Do not use
+        /// this with `deinit` or `add` as it will cause illegal behavior, use
+        /// `addAssumeCapacity` instead.
+        fn initBuffer(stops: []Stop) List {
```

Comment:
This is unexported and we don't use it, so we could probably remove it.

Reply:
Removed - just noting that we might need this eventually for testing or what not but it's trivial to add back if we do.